File: server/src/modules/assessment/repository_operations/update_assessment.rs
use chrono::Utc;
use sea_orm::*;
use uuid::Uuid;

use crate::utils::{AppError, AppResult};
use ::entity::assessments;

pub async fn update_assessment(
    db: &DatabaseConnection,
    id: Uuid,
    title: Option<String>,
    description: Option<String>,
    time_limit_minutes: Option<i32>,
    open_at: Option<chrono::NaiveDateTime>,
    close_at: Option<chrono::NaiveDateTime>,
    show_results_immediately: Option<bool>,
    term_number: Option<Option<i32>>,
    component: Option<Option<String>>,
    tos_id: Option<Option<String>>,
) -> AppResult<assessments::Model> {
    let tos_id = match tos_id {
        Some(Some(s)) if !s.is_empty() => {
            Some(Some(Uuid::parse_str(&s).map_err(|e| {
                AppError::BadRequest(format!("Invalid tos_id UUID: {}", e))
            })?))
        }
        Some(Some(_)) => Some(None),
        Some(None) => Some(None),
        None => None,
    };
    let mut assessment: assessments::ActiveModel = assessments::Entity::find_by_id(id)
        .one(db)
        .await
        .map_err(|e| AppError::InternalServerError(format!("Database error: {}", e)))?
        .ok_or_else(|| AppError::NotFound("Assessment not found".to_string()))?
        .into();

    if let Some(title) = title {
        assessment.title = Set(title);
    }
    if let Some(desc) = description {
        assessment.description = Set(Some(desc));
    }
    if let Some(time) = time_limit_minutes {
        assessment.time_limit_minutes = Set(time);
    }
    if let Some(open) = open_at {
        assessment.open_at = Set(open);
    }
    if let Some(close) = close_at {
        assessment.close_at = Set(close);
    }
    if let Some(show) = show_results_immediately {
        assessment.show_results_immediately = Set(show);
    }
    if let Some(q) = term_number {
        assessment.term_number = Set(q);
    }
    if let Some(c) = component {
        assessment.component = Set(c);
    }
    if let Some(tos) = tos_id {
        assessment.tos_id = Set(tos);
    }
    assessment.updated_at = Set(Utc::now().naive_utc());

    assessment
        .update(db)
        .await
        .map_err(|e| AppError::InternalServerError(format!("Failed to update assessment: {}", e)))
}

File: client/lib/domain/assessments/usecases/update_assessment.dart
import 'package:likha/core/sync/mutation_result.dart';
import 'package:likha/core/utils/typedef.dart';
import 'package:likha/domain/assessments/entities/assessment.dart';
import 'package:likha/domain/assessments/repositories/assessment_repository.dart';

class UpdateAssessmentParams {
  final String assessmentId;
  final String? title;
  final String? description;
  final int? timeLimitMinutes;
  final String? openAt;
  final String? closeAt;
  final bool? showResultsImmediately;
  final int? termNumber;
  final String? component;

  UpdateAssessmentParams({
    required this.assessmentId,
    this.title,
    this.description,
    this.timeLimitMinutes,
    this.openAt,
    this.closeAt,
    this.showResultsImmediately,
    this.termNumber,
    this.component,
  });
}

class UpdateAssessment {
  final AssessmentRepository _repository;

  UpdateAssessment(this._repository);

  ResultFuture<MutationResult<Assessment>> call(UpdateAssessmentParams params) {
    return _repository.updateAssessment(
      assessmentId: params.assessmentId,
      title: params.title,
      description: params.description,
      timeLimitMinutes: params.timeLimitMinutes,
      openAt: params.openAt,
      closeAt: params.closeAt,
      showResultsImmediately: params.showResultsImmediately,
      termNumber: params.termNumber,
      component: params.component,
    );
  }
}
